src/web/releases.rs

```diff
@@ -215,84 +216,84 @@ fn get_releases_by_owner(
-            releases.version AS version,
-            releases.description AS description,
-            releases.target_name AS target_name,
-            releases.release_time AS release_time,
-            releases.rustdoc_status AS rustdoc_status,
-            github_repos.stars AS github_stars,
-            COUNT(*) OVER() as total
-        FROM crates
-        INNER JOIN (
-            SELECT releases.id, releases.crate_id
-            FROM (
-                SELECT
-                    releases.id,
-                    releases.crate_id,
-                    RANK() OVER (PARTITION BY crate_id ORDER BY release_time DESC) as rank
-                FROM releases
-                WHERE releases.rustdoc_status AND NOT releases.yanked
-            ) AS releases
-            WHERE releases.rank = 1
-        ) AS latest_release ON latest_release.crate_id = crates.id
-        INNER JOIN releases ON latest_release.id = releases.id
-        LEFT JOIN github_repos ON releases.github_repo = github_repos.id
-        WHERE
-            ((char_length($1)::float - levenshtein(crates.name, $1)::float) / char_length($1)::float) >= 0.65
-            OR crates.name ILIKE CONCAT('%', $1, '%')
-        GROUP BY crates.id, releases.id, github_repos.stars
-        ORDER BY
-            levenshtein(crates.name, $1) ASC,
-            crates.name ILIKE CONCAT('%', $1, '%'),
-            releases.downloads DESC
-        LIMIT $2 OFFSET $3";
-
-    let rows = conn.query(statement, &[&query, &limit, &offset])?;
 
-    // Each row contains the total number of possible/valid results, just get it once
-    let total_results = rows
-        .get(0)
-        .map(|row| row.get::<_, i64>("total"))
-        .unwrap_or_default();
-    let packages: Vec<Release> = rows
+    use crate::utils::APP_USER_AGENT;
+    use once_cell::sync::Lazy;
+    use reqwest::blocking::Client as HttpClient;
+    use reqwest::header::{HeaderMap, HeaderValue, ACCEPT, USER_AGENT};
+
+    static HTTP_CLIENT: Lazy<HttpClient> = Lazy::new(|| {
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static(APP_USER_AGENT));
+        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
+        HttpClient::builder()
+            .default_headers(headers)
+            .build()
+            .unwrap()
+    });
+
+    let url = format!(
+        "https://crates.io/api/v1/crates?page={page}&per_page={limit}&q={query}",
+        page = page,
+        limit = limit,
+        query = query
+    );
+    debug!("fetching search results from {}", url);
+    let releases: CratesIoReleases = HTTP_CLIENT.get(&url).send()?.json()?;
+    let query = conn.prepare(
+        "SELECT github_repos.stars, releases.target_name, releases.rustdoc_status
+        FROM crates INNER JOIN releases ON crates.id = releases.crate_id
+                    LEFT JOIN github_repos ON releases.github_repo = github_repos.id
+        WHERE crates.name = $1 AND releases.version = $2",
```

Comment:
I don't quite understand - how can I query 30 crates at once in postgres? If I did something like `SELECT WHERE name in $1 and version in $2` then I wouldn't have the original data from crates.io, like `max_version`.

Reply:
a correct query would be something like 

```sql
select 
    ...
FROM 
    ... 
WHERE 
    (name = $1 and version = $2) OR 
    (name = $3 and version = $4) ... 
```
your query with 2x `IN` would fetch more records than needed. 

But yes, that would mean after you fetched the result into `releases`, there have to be multiple loops over the result: 
1. find our which additional data to fetch and build the postres query 
2. to merge `releases` and the db result . 
In my experience, the additional network roundtrips are far more expensive than additional in-memory loops. 

That being said, you could also say it's fast enough and keep it like it is :) 
I have a tendency to be a little more on _directly write faster features_ side and perhaps not enough on _delivering good enough results_ . There is always time to optimize later.

Reply:
when the could should stay readable, IMHO it's even fine to just fetch all versions for the crates in the result and filter them later.

Reply:
You can use an array parameter to pass all the versions to lookup at once: https://paste.ubuntu.com/p/QwtM8DRgJV/

Preferably that would be using a record to test each element (`(crates.name, releases.version) = ANY($1)`), but the `postgres` crate doesn't support records yet (https://github.com/sfackler/rust-postgres/issues/310), so I just serialized the name and version together into a string with a `:` separator which cannot appear in either value.

Reply:
but that `WHERE` on the concatenated value would not use any index, only if you create a special one, or write it into a field.

Reply:
A quick fix for that, prefilter to just the crates we're interested in:

```sql
WHERE crates.name = ANY($1)
AND crates.name || ':' || releases.version = ANY($2);
```

That goes directly through the `releases_crate_id_version_key` index on `releases` and just filters it by the specific versions required.

Reply:
yes, that would work :)

Reply:
@nemo157 yours has a slight difference, which is that it uses the description/release_time from our database instead of crates.io. But those fields should always be the same so that's seems fine if it lets us do everything in one database query. I'll use that, thanks!

Reply:
Yeah, it'd be _possible_ to thread those through from the API and match stuff up again afterwards, but I didn't see any reason to since grabbing it from the DB too seems cheap enough.